website/plugins/detect-broken-links/index.js

```diff
@@ -0,0 +1,209 @@
+// this is mostly taken from
+// https://github.com/trevorblades/gatsby-remark-check-links
+
+const visit = require("unist-util-visit");
+
+module.exports = async function plugin({
+  markdownAST,
+  markdownNode,
+  cache,
+  getCache,
+  getNode,
+  files,
+}) {
+  // if there's no slug, it is not a markdown file we're interested in
+  if (!markdownNode.fields?.slug) {
+    return markdownAST;
+  }
+
+  const headingAnchors = [];
+  const links = [];
+
+  // collect all links and headings from the current file
+  visit(markdownAST, "link", (node, _, parent) => {
+    // if this is the little link anchor next to a heading
+    // note: this method of header discovery assumes that every header
+    //       is covered by gatsby-remark-autolink-headers
+    if (parent.type === "heading") {
+      // the id is the generated link for this heading
+      headingAnchors.push(parent.data.id);
+      return;
+    }
+
+    const isHttpLink = /^https?\:\/\//.test(node.url);
+
+    // if this is a link that does not point to one of our documents
+    if (isHttpLink) {
+      return;
+    }
+
+    links.push({
+      position: node.position,
+      url: node.url,
+      frontmatter: markdownNode.frontmatter,
+    });
+  });
+
+  // collect all links from definitions (non-inline links)
+  visit(markdownAST, "definition", (node, _, parent) => {
+    const isHttpLink = /^https?\:\/\//.test(node.url);
+
+    // if this is a link that does not point to one of our documents
+    if (isHttpLink) {
+      return;
+    }
+
+    links.push({
+      position: node.position,
+      url: node.url,
+      frontmatter: markdownNode.frontmatter,
+    });
+  });
+
+  const parent = await getNode(markdownNode.parent);
+
+  cache.set(getCacheKey(parent), {
+    slug: markdownNode.fields.slug,
+    links,
+    headingAnchors: headingAnchors,
+    absolutePath: parent.absolutePath,
+  });
+
+  const documentMap = {};
+
+  // check if all other files have entries in the cache
+  // and if they do populate the local cache above
+  for (let index = 0; index < files.length; index++) {
+    const file = files[index];
+
+    // ignore non mdx files
+    if (!/^mdx?$/.test(file.extension)) {
+      continue;
+    }
+
+    const cacheKey = getCacheKey(file);
+
+    let cacheEntry = await cache.get(cacheKey);
+    if (!cacheEntry && getCache) {
+      // the cache provided to `gatsby-mdx` has its own namespace, and it
+      // doesn't have access to `getCache`, so we have to check to see if
+      // those files have been visited here.
+      const mdxCache = getCache("gatsby-plugin-mdx");
+      cacheEntry = await mdxCache.get(cacheKey);
+    }
+
+    if (cacheEntry) {
+      documentMap[cacheEntry.slug] = cacheEntry;
+
+      continue;
+    }
+
+    // don't continue if a page hasn't been visited yet
+    return markdownAST;
+  }
```

Comment:
@rstaib I don't really like this part. Essentially it checks if all markdown files had their links/headings added to the cache:
- If not, exit
- If they have, continue
This means the loop checks every file for every file emitted... The original plugin was doing it like that and I couldn't find a more suitable spot to put the logic after all files have been checked.
There are additional events like `onPostBuild`, etc. but in them I would have to be able to access all of the files and get their cache entries (essentially what is happening in the loop, but only executed once). Any ideas?